File: src/ReplicatedStorage/Scripts/Drain.server.lua
local drain = script.Parent
local utils = require(game.ServerScriptService.Utils)
local pinball = utils.getParentPinball(drain)
local values = pinball:FindFirstChild("Values")

drain.Touched:Connect(function(part)
	if part.Parent.Name ~= "Balls" then
		return
	end
	if values.OwnerValue.Value ~= nil then
		local player = values.OwnerValue.Value
		player.Character.Humanoid.Health = 0
		local leaderstats = player:FindFirstChild("leaderstats")
		if leaderstats ~= nil then
			local score = leaderstats:FindFirstChild("Score")
			if score ~= nil then
				score.Value = 0
			end
		end
	end
	task.wait(0.5)
	part:Destroy()
end)

File: src/ServerScriptService/components/Drain.server.ts
import { BaseComponent, Component } from '@flamework/components'
import { OnStart } from '@flamework/core'
import { store } from 'ServerScriptService/store'
import {
  getArcadeTableOwner,
  getParentArcadeTable,
} from 'ServerScriptService/utils'

@Component({ tag: 'Drain' })
export class DrainComponent extends BaseComponent implements OnStart {
  onStart() {
    const drain = this.instance as BasePart
    const arcadeTable = getParentArcadeTable(this.instance)
    if (!arcadeTable) return

    drain.Touched?.Connect((part) => {
      if (part.Parent?.Name !== 'Balls') return

      const player = getArcadeTableOwner(arcadeTable)
      if (player) {
        store.resetScore(player.UserId)
        const character: (Model & { Humanoid?: Humanoid }) | undefined =
          player.Character
        if (character?.Humanoid) character.Humanoid.Health = 0
      }

      task.wait(0.5)
      part.Destroy()
    })
  }
}
